integration/instance.go

```diff
@@ -0,0 +1,392 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/purpleidea/mgmt/recwatch"
+
+	errwrap "github.com/pkg/errors"
+)
+
+const (
+	// RootDirectory is the directory that is exposed in the per instance
+	// directory which can be used by that instance safely.
+	RootDirectory = "root"
+
+	// PrefixDirectory is the directory that is exposed in the per instance
+	// directory which is used for the mgmt prefix.
+	PrefixDirectory = "prefix"
+
+	// ConvergedStatusFile is the name of the file which is used for the
+	// converged status tracking.
+	ConvergedStatusFile = "csf.txt"
+
+	// StdoutStderrFile is the name of the file which is used for the
+	// command output.
+	StdoutStderrFile = "stdoutstderr.txt"
+
+	// longTimeout is a high bound of time we're willing to wait for events.
+	// If we exceed this timeout, then it's likely we are blocked somewhere.
+	longTimeout = 30 // seconds
+
+	// convergedTimeout is the number of seconds we wait for our instance to
+	// remain unchanged to be considered as converged.
+	convergedTimeout = 5 // seconds
+
+	// dirMode is the the mode used when making directories.
+	dirMode = 0755
+
+	// fileMode is the the mode used when making files.
+	fileMode = 0644
+)
+
+// Instance represents a single running mgmt instance. It is a building block
+// that can be used to run standalone tests, or combined to run clustered tests.
+type Instance struct {
+	// Hostname is a unique identifier for this instance.
+	Hostname string
+
+	// Preserve prevents the runtime output from being explicitly deleted.
+	// This is helpful for running analysis or tests on the output.
+	Preserve bool
```

Comment:
Like said before I think there is no reason why this should be an Instance attribute. Now you have to pass it around everywhere. Maybe better to always preserve on error and set this as a testsuite constant or a test flag instead so it can be enabled if you really want to see the output of a passed test?